refactor(*): implement a new design for interfaces

blackHatMonkey opened the pull request:
Implement a new design for interfaces module where getter and setters
are removed. Some of the methods have been removed or replaced to
make using network manager class much easier. Error messages are
simplified as well. Most importantly a tests suite is written for
the module and the coverage is 100%.

sophron commented:
Nice, properties and unit-testing. I'll have a look and possibly merge soon.

sophron commented on wifiphisher/common/interfaces.py:
```diff
@@ -217,180 +272,220 @@ def __init__(self):
+        # populate our dictionary with all the available interfaces on the system
+        for interface in pyw.interfaces():
+            try:
+                card = pyw.getcard(interface)
```
This seems to throw a `pyric.error` (Errno 19: No such device)

The problem is that that `pyw.interfaces` returns all interfaces in the system including loopback but there's no device card tied to the loopback interface.

blackHatMonkey replied:
I didn't catch that error during my tests. I will add the exception.

sophron commented on wifiphisher/common/interfaces.py:
```diff
@@ -1,163 +1,147 @@
-class NotEnoughInterfacesFoundError(Exception):
-    """
-    Exception class to raise in case of a finding less than enough interfaces.
-    """
```
Why is that removed? I don't see where you check the number of available interfaces in the system in the new code.

blackHatMonkey replied:
Checking for the number of interfaces in not needed as I demonstrate with this example:

Cards:
- AP

This will throw an error stating that an interface with monitor mode is required but no found.

Cards:
- AP
- MON

There is no problem and would continue as normal.

Cards:
- MON


This will throw an error stating that an interface with AP mode is required but no found.

sophron replied:
I think this will raise confusion. If a user has one card that supports both AP and Monitor modes, he will see an error that he needs an interface that supports Monitor Mode. It won't be clear to the user that he needs a separate card for Monitor Mode.

We need to be clear when raising an exception. Not having enough adapters is different from having the required number of adapters but they are not supporting the right modes. Otherwise, people will start opening issues about what is required again.

blackHatMonkey replied:
Imagine this scenario where I have a wired interface and a wireless interface. It passes the two interface test while throwing the same error.

sophron replied:
Of'course the check would be if there are two wireless interfaces (`pyric.iswireless`) installed.

blackHatMonkey replied:
I think the best option would be to reword this to include a message regarding the number of interfaces.

sophron commented on wifiphisher/common/interfaces.py:
```diff
@@ -217,180 +272,220 @@ def __init__(self):
+        """
+
+        monitor_interface = self.get_interface(has_monitor_mode=True)
+        ap_interface = self.get_interface(has_ap_mode=True)
```
What is the problem with our current algorithm that automatically picks the right interfaces?

The new code doesn't look correct. Imagine the following scenario:

wlan0: supports monitor and ap mode
wlan1: supports monitor mode only

In that case, `ap_interface` should get wlan0 and `monitor_interface` should get `wlan1` but with this code they won't be assigned like that.

blackHatMonkey replied:
I thought about it myself but it seems that I have messed up the order. Using your example swapping them should work since if interface has AP it is guaranteed to have monitor mode as well.

sophron replied:
>  if interface has AP it is guaranteed to have monitor mode as well

I'm not sure that's correct. What's your source?

blackHatMonkey replied:
I will try to come up with a better algorithm to fix this.

sophron commented on wifiphisher/pywifiphisher.py:
```diff
@@ -569,52 +575,55 @@ def start(self):
+
+            # make sure interfaces are not blocked
+            self.network_manager.unblock_interface(ap_iface)
+            self.network_manager.unblock_interface(mon_iface)
```
In various places, we also need to make sure that the cards are not invalid. See my comment on #580.

sophron commented on tests/test_interfaces.py:
```diff
@@ -0,0 +1,595 @@
+
+    def test_is_interface_valid_no_ap_error(self):
+        """
+        Tests is_interface_valid method when interface is has no AP
```
* when interface has

sophron commented on tests/test_interfaces.py:
```diff
@@ -0,0 +1,595 @@
+
+    def test_is_interface_valid_has_ap_none(self):
+        """
+        Tests is_interface_valid method when interface is has AP
```
* when interface has

This typo is in many places.

blackHatMonkey replied:
Thats what you get for copy-pasting :cry: .

sophron commented on tests/test_interfaces.py:
```diff
@@ -0,0 +1,595 @@
+        """
+        Tests get_interface_automatically method when two interfaces
+        can be found
+        """
```
This is one of the most important tests. But you only examine one simple case.

You should examine all possible weird corner cases.

blackHatMonkey replied:
I will add more tests in there.

sophron commented on tests/test_interfaces.py:
```diff
@@ -0,0 +1,595 @@
+# pylint: skip-file
+""" This module tests the interface module """
```
I'm not sure I understand how you are asserting? I see some methods included from Pyric. Or how you do you mock the interfaces? Whatever you do, please document it.  :)

blackHatMonkey replied:
I don't understand what you mean. I just check that the code calls `PyRIC`'s function and in some cases change the return values for some functions.

blackHatMonkey commented:
@sophron @anakin1028 I have made some changes to this PR. Can you please review it and let me know if any changes need to be made.

Disclaimer: I haven't had the chance to test it so it might not work.

anakin1028 commented on wifiphisher/common/interfaces.py:
```diff
@@ -319,233 +339,320 @@ def __init__(self):
+        self.set_interface_mode(interface_name, "managed")
+
+        try:
+            pyw.macset(card, mac_address)
```
Before setting mac address we should check whether the card is down. This is my first error to executing on this branch. I will fixed it on my local and see if there are any problems else.

blackHatMonkey replied:
is `pyw.up(card)` enough?

anakin1028 replied:
also checks if the card is in `managed` mode

blackHatMonkey replied:
>Before setting mac address we should check whether the card is down. This is my first error to executing on this branch. I will fixed it on my local and see if there are any problems else.

Since I'm setting the card to `managed on line 395 the interface should b e 100% up.

>also checks if the card is in managed mode

On line 395 I'm setting it to `managed`.

anakin1028 replied:
@blackHatMonkey , yes. But mac address can only change when the card is in down state.

blackHatMonkey replied:
Did not realize this. I will make the changes and document it :smile.

anakin1028 replied:
OK cheers :)

anakin1028 commented on wifiphisher/pywifiphisher.py:
```diff
@@ -586,69 +590,73 @@ def start(self):
             if not args.nojamming:
-                print ("[{0}+{1}] " + mon_iface.get_name() + ' mac address becomes '
-                     + mon_iface.get_current_mac()).format(G, W)
+                print ("[{0}+{1}] {2} mac address becomes {3}".format(G, W, mon_iface))
```
you don't give the mac address argument

anakin1028 replied:
line#669, `up_ifaces` is not defined in `NetworkManager`

anakin1028 replied:
in `select_access_point` you should pass the `NetworkManager` inside.

anakin1028 replied:
check this statement `self.access_point.set_interface(ap_iface.get_name())`. `ap_iface` is already a string.

blackHatMonkey commented:
@anakin1028 How did you see the mac randomization? Do you think we can do it better?

anakin1028 commented:
Brian, after the problems I have mentioned are fixed, I'll do the functional test :)

anakin1028 commented on wifiphisher/common/interfaces.py:
```diff
@@ -319,233 +339,320 @@ def __init__(self):
-        self.ap_iface = ""
-        self.jam_iface = ""
+        self._name_to_object = dict()
+        self._active = set()
```
is the name `_name_to_object` clear enough to represent it is a dictionary doing mapping from adapter name to the `NetworkAdapter`?

blackHatMonkey replied:
do you have something else in mind?

anakin1028 commented on wifiphisher/common/interfaces.py:
```diff
@@ -319,233 +339,320 @@ def __init__(self):
-        for i in ifaces:
-            card = pyw.getcard(i.get_name())
-            pyw.up(card)
-    
```
why `up_ifaces` removed? I remember in `pywifiphisher.py` will call this function, or you have refactored this part too?

blackHatMonkey replied:
yes we no longer need this function. I just forgot to remove it :cry: .

sophron commented on tests/test_interfaces.py:
```diff
@@ -0,0 +1,779 @@
+import mock
+import wifiphisher.common.interfaces as interfaces
+import pyric
+
```
You will have to set your path accordingly to execute the tests from whatever location. See how I did it in my tests.

blackHatMonkey replied:
instead of doing that for all of our tests we should use nose or some similar tool. All you have to do is run nose from the `wifiphisher` directory.

blackHatMonkey replied:
Or maybe we can leverage `tox`?

sophron commented on tests/test_interfaces.py:
```diff
@@ -0,0 +1,779 @@
+class TestNetworkAdapter(unittest.TestCase):
+    """ Tests NetworkAdapter class """
+
+    @mock.patch("wifiphisher.common.interfaces.pyw")
```
What is the reason for mocking here?

blackHatMonkey replied:
I believe it was because of `pyw.card`. However we don't need it and I will remove it.

sophron commented on tests/test_interfaces.py:
```diff
@@ -0,0 +1,779 @@
+        adapter.is_wireless = False
+        self.network_manager._name_to_object[interface_name] = adapter
+
+        self.network_manager.is_interface_wired(interface_name)
```
I don't see any `assert` here.

sophron commented:
Brian,

A few notes on your tests.

In general, unit tests should check the behavior of the code that you have written. The test `test_has_monitor_mode_true` that checks if a property of an object has been set properly to True, is a unit test in Python implementation itself and not in your code. If, however, there was a logic that you have written behind setting the property (e.g. the monitor interface would be true in specific cases) then it would make sense to write a test for it.

Same applies for most of your tests. Tests that check if the properties have been set properly (e.g. `test_mac_address_original` and `test_mac_address_modified`) are not offering something. A test however that would check the `recover_mac_address_to_original` method if it did recover the mac address to its previous state would make sense.

All in all, what we want to test is the logic of the tool, not if Python sets the properties correctly. The first case would break if the application logic changes, the second type would break simply if we rename a field. It's important to have specific, not-useless test cases that we can easily maintain and people would get the idea of a module simply by reading them.

blackHatMonkey commented:
@sophron @anakin1028 I have fixed all the problems you mentioned. I have done a basic test and it should work now but it requires a full testing.

@sophron about the `test_has_monitor_mode_true` I'm checking the property and not the actual variable. So even if we change the variable it should still work. Also I have fixed a bunch other cases where I was checking for variables :smile:.

blackHatMonkey commented:
@anakin1028 By `doing better` I meant that right now we are doing randomization in two places. I think we should come up with a better plan so that we only have to do it once.

sophron commented:
Nice stuff!

@anakin1028 feel free to merge this if you have no other comments.

anakin1028 commented:
@blackHatMonkey Nice~~~ I just help fixed one small problem, that is in `pywifiphisher` the `mon_iface` variable is string, but you still use it to set the channel number. otherwise it looks good, I'll merge it to master :)
